pkg/diag/validator/pod.go

```diff
@@ -120,37 +120,80 @@ func (p *PodValidator) getPodStatus(pod *v1.Pod) *podStatus {
-				// TODO(dgageot): Add EphemeralContainerStatuses
-				cs := append(pod.Status.InitContainerStatuses, pod.Status.ContainerStatuses...)
-				// See https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#container-states
-				statusCode, logs, err := getContainerStatus(pod, cs)
-				if statusCode == proto.StatusCode_STATUSCHECK_POD_INITIALIZING {
-					// Determine if an init container is still running and fetch the init logs.
-					for _, c := range pod.Status.InitContainerStatuses {
-						if c.State.Waiting != nil {
-							return statusCode, []string{}, fmt.Errorf("waiting for init container %s to start", c.Name)
-						} else if c.State.Running != nil {
-							return statusCode, getPodLogs(pod, c.Name), fmt.Errorf("waiting for init container %s to complete", c.Name)
-						}
-					}
+	if isPodReady(pod) {
+		return proto.StatusCode_STATUSCHECK_SUCCESS, nil, nil
+	}
+	if c, ok := isPodNotScheduled(pod); ok {
+		logrus.Debugf("Pod %q not scheduled: checking tolerations", pod.Name)
+		sc, err := getUntoleratedTaints(c.Reason, c.Message)
+		return sc, nil, err
+	}
+
+	if isPodScheduledButNotReady(pod) {
+		logrus.Debugf("Pod %q scheduled but not ready: checking container statuses", pod.Name)
+		// TODO(dgageot): Add EphemeralContainerStatuses
+		cs := append(pod.Status.InitContainerStatuses, pod.Status.ContainerStatuses...)
+		// See https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#container-states
+		statusCode, logs, err := getContainerStatus(pod, cs)
+		if statusCode == proto.StatusCode_STATUSCHECK_POD_INITIALIZING {
+			// Determine if an init container is still running and fetch the init logs.
+			for _, c := range pod.Status.InitContainerStatuses {
+				if c.State.Waiting != nil {
+					return statusCode, []string{}, fmt.Errorf("waiting for init container %s to start", c.Name)
+				} else if c.State.Running != nil {
+					return statusCode, getPodLogs(pod, c.Name), fmt.Errorf("waiting for init container %s to complete", c.Name)
 				}
-				return statusCode, logs, err
-			case v1.ConditionUnknown:
-				logrus.Debugf("Pod %q scheduling condition is unknown", pod.Name)
-				return proto.StatusCode_STATUSCHECK_UNKNOWN, nil, fmt.Errorf(c.Message)
 			}
 		}
+		return statusCode, logs, err
 	}
-	return proto.StatusCode_STATUSCHECK_SUCCESS, nil, nil
+
+	if c, ok := isPodStatusUnknown(pod); ok {
+		logrus.Debugf("Pod %q condition status of type %s is unknown", pod.Name, c.Type)
+		return proto.StatusCode_STATUSCHECK_UNKNOWN, nil, fmt.Errorf(c.Message)
+	}
+
+	logrus.Debugf("Unable to determine current service state of pod %q", pod.Name)
+	return proto.StatusCode_STATUSCHECK_UNKNOWN, nil, fmt.Errorf("unable to determine current service state of pod %q", pod.Name)
+}
+
+func isPodReady(pod *v1.Pod) bool {
+	for _, c := range pod.Status.Conditions {
+		if c.Type == v1.ContainersReady && c.Status == v1.ConditionTrue {
+			return true
+		}
+	}
+	return false
```

Comment:
Use the pod's Ready status: the containers could be ready, but the pod itself might not yet be.

Reply:
done.